aidial_adapter_vertexai/embedding/multi_modal.py

```diff
@@ -0,0 +1,224 @@
+from logging import DEBUG
+from typing import AsyncIterator, List, Tuple
+
+from aidial_sdk.chat_completion.request import Attachment
+from aidial_sdk.embeddings import Response as EmbeddingsResponse
+from aidial_sdk.embeddings import Usage
+from aidial_sdk.embeddings.request import EmbeddingsRequest
+from pydantic import BaseModel
+from vertexai.vision_models import (
+    Image,
+    MultiModalEmbeddingModel,
+    MultiModalEmbeddingResponse,
+)
+
+from aidial_adapter_vertexai.chat.errors import UserError, ValidationError
+from aidial_adapter_vertexai.dial_api.attachments import (
+    derive_attachment_mime_type,
+    download_attachment,
+)
+from aidial_adapter_vertexai.dial_api.embedding_inputs import (
+    EMPTY_INPUT_LIST_ERROR,
+    collect_embedding_inputs,
+)
+from aidial_adapter_vertexai.dial_api.response import make_embeddings_response
+from aidial_adapter_vertexai.dial_api.storage import (
+    FileStorage,
+    create_file_storage,
+)
+from aidial_adapter_vertexai.embedding.embeddings_adapter import (
+    EmbeddingsAdapter,
+)
+from aidial_adapter_vertexai.embedding.encoding import vector_to_base64
+from aidial_adapter_vertexai.utils.json import json_dumps_short
+from aidial_adapter_vertexai.utils.log_config import vertex_ai_logger as log
+from aidial_adapter_vertexai.vertex_ai import (
+    get_multi_modal_embedding_model,
+    init_vertex_ai,
+)
+
+# See the documentation: https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/multimodal-embeddings-api
+
+SUPPORTED_IMAGE_TYPES = ["image/jpeg", "image/png"]
+
+
+class ModelRequest(BaseModel):
+    class Config:
+        arbitrary_types_allowed = True
+
+    image: Image | None = None
+    contextual_text: str | None = None
+
+    def extract_embeddings(
+        self, response: MultiModalEmbeddingResponse
+    ) -> Tuple[List[float], int]:
+        # The model doesn't report the number of input tokens
+        tokens = 1
+
+        vector: List[float] | None = None
+        if self.image:
+            vector = response.image_embedding
+        else:
+            vector = response.text_embedding
+
+        if vector is None:
+            raise ValueError("No embeddings returned")
+
+        return vector, tokens
+
+
+async def compute_embeddings(
+    request: ModelRequest,
+    model: MultiModalEmbeddingModel,
+    dimensions: int | None,
+) -> Tuple[List[float], int]:
+
+    if log.isEnabledFor(DEBUG):
+        msg = json_dumps_short(
+            {
+                "image": request.image,
+                "contextual_text": request.contextual_text,
+                "dimension": dimensions,
+            }
+        )
+        log.debug(f"request: {msg}")
+
+    response: MultiModalEmbeddingResponse = model.get_embeddings(
```

Comment:
Unfortunately, under the hood model is using `urllib3` for requests, and this would block whole event loop of application


So we need to run this in executor

Reply:
Indeed, fixed